### PR TITLE
Performance improvements parsing UnverifiedJsonWebToken

### PR DESCRIPTION
**Pull request by carterkozak**
Implemented JMH benchmarks for UnverifiedJsonWebToken.tryParse.

Results before this change:
```
UnverifiedJsonWebTokenBenchmarks.parseNonJwt        thrpt    3  8530177.987 ± 438717.090  ops/s
UnverifiedJsonWebTokenBenchmarks.parseSessionToken  thrpt    3   502692.902 ±  19562.944  ops/s
```

With this change:
```
UnverifiedJsonWebTokenBenchmarks.parseNonJwt        thrpt    3  14073771.303 ± 1743291.960  ops/s
UnverifiedJsonWebTokenBenchmarks.parseSessionToken  thrpt    3    662978.619 ±   66605.517  ops/s
```

### CODE REVIEW

**Comment by markelliot** [comment]
We probably ought to rerun this with AfterBurner enabled.

**Comment by carterkozak** [comment]
I tested with afterburner, it didn't make much difference in the current implementation, however combined with a few other optimizations it can increase the `parseSessionToken` results from ~600k to ~900k ops/sec. I didn't push those because the implementation sacrifices readability. Happy to put those together if you're interested, probably works best in a separate PR.

**Comment by carterkozak** [comment]
Sample implementation, needs to be cleaned up before it can be merged, but it's consistently providing >1m ops/sec
https://github.com/palantir/auth-tokens/commit/3510e81847c9a8c31b2a632752db5d31b1f7e5bc

File: auth-tokens/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebToken.java [hunk 1]
```diff
@@ -44,9 +45,10 @@
 @ImmutablesStyle
 public abstract class UnverifiedJsonWebToken {
 
-    private static final ObjectMapper MAPPER = new ObjectMapper()
+    private static final ObjectReader READER = new ObjectMapper()
             .registerModule(new Jdk8Module())
-            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
```
Comment (ferozco) [inline auth-tokens/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebToken.java:48]: why do we ignore unknown properties?

File: benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+@Fork(1)
+@Threads(1)
+@Warmup(iterations = 10, time = 5)
+@Measurement(iterations = 3, time = 3)
+public class UnverifiedJsonWebTokenBenchmarks {
```
Comment (ferozco) [inline benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java:33]: Where do we plan on using these benchmarks?

**Review by ferozco — COMMENTED** [review]
Just out of curiosity, what motivated this change?

File: benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java (hunk 2 above)
Comment (carterkozak) [reply benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java:33]: We don't have a great story for benchmark automation at the moment, so running benchmarks is a manual action.

File: auth-tokens/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebToken.java (hunk 1 above)
Comment (carterkozak) [reply auth-tokens/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebToken.java:48]: Allows us to add additional fields to the jwt payload in the future without breaking existing consumers.

File: benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        return UnverifiedJsonWebToken.tryParse(NOT_JWT);
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public final Optional<UnverifiedJsonWebToken> parseSessionToken() {
+        return UnverifiedJsonWebToken.tryParse(SESSION_TOKEN);
+    }
+
+    private static String repeat(CharSequence seq, int count) {
```
Comment (ellisjoe) [inline benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java]: can use `Strings.repeat(...)` from guava

**Review by ellisjoe — COMMENTED** [review]
one minor nit

File: benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java (hunk 3 above)
Comment (carterkozak) [reply benchmarks/src/main/java/com/palantir/tokens/auth/UnverifiedJsonWebTokenBenchmarks.java]: We don't have guava in this module. I can add the dependency, but this was simpler.

**Review by ellisjoe — APPROVED** [review]

**Pull request merged**